### ISSUE TITLE
Add 'pulumi state protect' command

### ISSUE THREAD

**Issue by gpduck** [pulumi/pulumi#6242]
We currently occasionally have to unprotect resources for troubleshooting purposes due to #5749 and it would be nice if there was a way to re-protect them from the command line. I can't really think of any use cases outside of this scenario, but it still seems appropriate to have a way to "undo" the unprotect operation from the command line as well.

**Comment by clstokes** [comment]
@gpduck Would the exact same semantics as https://www.pulumi.com/docs/reference/cli/pulumi_state_unprotect/ work here?

e.g.
`pulumi state protect <resource URN>`
and 
`pulumi state protect --all`

**Comment by gpduck** [comment]
@clstokes yeah I think that seems fair. It would let people undo the unprotect operations easily by just changing the command name.

**Comment by stack72** [comment]
Hi @gpduck 

So I was just looking at this issue - so the reason state unprotect works is that it would mean a delete operation will work without first needing to run an update

With a state protect command, the resources themselves would need to be updated with the 'protect: true' resource option or the next Pulumi update would effectively remove the protection again

This seems like it would negate the usefulness of the state protect command

I'd love your thoughts here

Paul

**Comment by gpduck** [comment]
In the case mentioned in #5749 we're unprotecting so that we can get a diff, we don't actually want to delete the resources at all, so running a pulumi up after we do unprotect to get the protection re-enabled is extra scary.

**Comment by komalali** [comment]
I'm curious if we've considered fixing this by addressing https://github.com/pulumi/pulumi/issues/5749 to generate a diff for protected resources rather than adding a new command for a very niche use-case?

**Comment by leezen** [comment]
> In the case mentioned in #5749 we're unprotecting so that we can get a diff, we don't actually want to delete the resources at all, so running a pulumi up after we do unprotect to get the protection re-enabled is extra scary.

I have a slightly different take on this and am curious if the following would work for you @gpduck. If I understand the use case correctly, is the following true?

1. You are at commit A with a corresponding state (S) that matches the code at this commit.
2. You write some code (commit B) and want to see the diff, which results in hitting #5749.
3. You unprotect the resources (generating state S') in order to see the diff.
4. At this point, you'd like to re-protect the resources.

I understand the hesitancy to run `pulumi up` with commit B since you don't actually want to replace the resources. But is there a reason you cannot checkout commit A and run `pulumi up`? I would expect the diff at that point to only be adding the `protect` flag?

**Comment by gpduck** [comment]
If the mismatch is actually due to code changes then re-running pulumi up on commit A would safely re-protect the resources (although we'd still be running it without the protect:true safety net). But there are also some times when the change has occurred in Azure and the preferred way forward is for us to update our pulumi code to match the current Azure state.

For example we recently had a team do a restore on an Azure VM, which ended up renaming the OS disk. The next time we ran the pulumi project it wanted to recreate the VM to "fix" the OS disk name. It turns out we don't super care about the name of the disk resource so we're updating our pulumi libraries to ignore those changes. But the only way to identify why the resource was being replaced was to unprotect it, and then the only way to re-protect it is to correctly update the project and have the protection re-applied (or to manually update state, which I try to avoid as much as possible).

Also just in general it is nice to have the corresponding commands to "undo" changes that people have manually made :)

**Comment by leezen** [comment]
@gpduck Thanks for the detailed explanation. Sounds like having the separate command would make more sense for you. Thanks!

**Comment by komalali** [comment]
I can see that there are scenarios where `undo unprotect` is a valuable option to have. My misgivings are around the semantics -  specifically with the choice of `pulumi state protect` as the command for `undo unprotect`. 

I feel like this could confuse people because `pulumi state protect` would actually mean nothing unless you:
1. Have resources that are protected in your code.
2. Have run `unprotect` on them.

As an alternative, I'd be more comfortable with an `--undo` flag on `unprotect` that would work exactly the way [described here](https://github.com/pulumi/pulumi/issues/6242#issuecomment-773527259) but using `pulumi state unprotect --undo` in place of `pulumi state protect`. I'll admit it's not the most ergonomic of choices, but I feel that it is a much more accurate representation of the actual functionality - and adding a flag with somewhat confusing semantics feels preferable to adding an entire command with confusing semantics and caveats.

Just my $0.02 🙏

**Comment by brandonpollack23** [comment]
I tend to agree with @komalali's point here.  I think reprotect is a good compromise and ill throw together a quick implementation for discussion.

Basically the name reprotect implies that you can only reprotect resources that were unprotected (and therefore were protected in teh IAC code itself in teh first place).  We can also explain as much in the docs of the command.

**Comment by brandonpollack23** [comment]
Some notes from a discussion:
There are dragons but there is not necessarily a state we can get it in that results in a "protected" value to not remain protect.

When you run your pulumi program an "up" can potentially unprotect it.

When we switch to destroy running the program it should be safe as in:
* https://github.com/pulumi/pulumi/issues/13428#issuecomment-1637602451

There are potential footguns, so perhaps we should have a warning with scary text in the docs and require a "--yes" or interactive yes.  It always prints scary warnings and let them know an `up` might remove the protection.

Our docs should also have scary warnings and link to the Resource Option for protect.

given all these we've agreed `protect` is a good name
